### PR TITLE
Add Queue [Java] (#509)

### LINKED ISSUE #509: Queue [Java]
#### This issue is about [Queue](http://www.cdn.geeksforgeeks.org/stack-data-structure-introduction-program/).

- [x] Issue Title is in form `Algo/DS Name [Language]`.
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This Algo/DS is worth adding to this repository.
- [x] Assign this issue to me (I want to work on this).
	- [x] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [x] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by sebasrp**
Fixes #509

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

**Comment by singhpratyush** [comment]
Hi! Thanks for showing interest in contributing to this repository.

Please make sure that you have ticked the points mentioned in PR description correctly.

Thanks again!

File: queue/Queue.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import java.util.NoSuchElementException;
```
Comment (Prakash2403) [inline queue/Queue.java]: Either import `java.util.Queue` or don't use `implements Queue<T>`. 

*Note*: If you decide to import `java.util.Queue`, then override `clear` method also.

Comment (sebasrp) [reply queue/Queue.java]: Agree - this was a leftover from work i was doing to implement queue in different ways. Removed

**Comment by sebasrp** [comment]
update README.md with a new row for Queue

**Comment by Prakash2403** [comment]
@sebasrp Nice work. Just one more change. Since, your `Queuet` class is public, either you have to change the filename to `Queuet.java` or you have to change classname to `Queue`. Please do that.

**Comment by sebasrp** [comment]
sorry - my mistake, the name was incorrectly set. fixing that.

**Comment by Prakash2403** [comment]
@sebasrp Hey, just one more change. You have to define `main` method in `Queue` class and add some tests. In case of doubt, refer to other codes and you will understand.

File: queue/Queue.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    public static void main(String[] args) {
+        Queue<Integer> queue = new Queue<>();
+        for(int i=1; i<=10; i++) // Creates a dummy queue which contains integers from 1-10
+        {
+            queue.add(i);
+        }
+
+        System.out.println("QUEUE :");
+
+        while (queue.peek != null)
```
Comment (Prakash2403) [inline queue/Queue.java]: it should be `queue.peek()`.

Comment (sebasrp) [reply queue/Queue.java]: sorry - pushed wrong version. fixing now

File: queue/Queue.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /**
+     * Retrieves and removes the head of this queue.
+     * This method differs from poll only in that it throws an exception if this queue is empty.
+     * @return the head of this queue.
+     * @throws NoSuchElementException if this queue is empty
+     */
+    public T remove() throws NoSuchElementException {
+        if (total == 0) throw new NoSuchElementException();
```
Comment (singhpratyush) [inline queue/Queue.java]: Please add a line break after control statements.

File: queue/Queue.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    public static void main(String[] args) {
+        Queue<Integer> queue = new Queue<>();
+        for(int i=1; i<=10; i++) // Creates a dummy queue which contains integers from 1-10
+        {
+            queue.add(i);
+        }
+
+        System.out.println("Queue :");
+
+        while (queue.peek() != null)
```
Comment (singhpratyush) [inline queue/Queue.java]: Please start braces in same line - 
```java
while (...) {
    ...
}
```

File: queue/Queue.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Retrieves and removes the head of this queue, or returns null if this queue is empty.
+     * This method differs from peek only in that it throws an exception if this queue is empty.
+     * @return the head of this queue, or returns null if this queue is empty.
+     */
+    public T poll() {
+        try{
+            return remove();
+        } catch (Exception ex){
```
Comment (singhpratyush) [inline queue/Queue.java]: Would be great if you could change `} catch (..){` to `} catch (...) {` so that it stays consistent everywhere.

**Comment by aashutoshrathi** [comment]
Resolve conflicts.

**Comment by singhpratyush** [comment]
@sebasrp: There is a conflict. Also, there are many unaddressed issues that need to be fixed. Please help us to get this resolved. Thanks.

**Comment by sebasrp** [comment]
will fix this over the weekend, i am currently traveling.
thanks

File: queue/Queue.java (hunk 3 above)
Comment (sebasrp) [reply queue/Queue.java]: done

File: queue/Queue.java (hunk 5 above)
Comment (sebasrp) [reply queue/Queue.java]: done (in previous catch statement as well)

File: queue/Queue.java (hunk 4 above)
Comment (sebasrp) [reply queue/Queue.java]: done

**Pull request merged**